### DISCUSSION TITLE
Importing Function Definitions from the OpenAI json schema

### DISCUSSION THREAD

**Discussion by raffaeler** [microsoft/semantic-kernel#8835] [Q&A] [answered]
Hello,
I have to share the function/tool definitions with another library written using the OpenAI library.
Instead of using attributes to build my plugins, I would like to import the data using the json schema that I use in the other library.

While I could create a C# code generator, I really hope there is a faster solution for this.

Can someone point me to the API, if exists?
Thanks!

**Comment by evchaki** [comment]
@raffaeler - you can take a look at how to do this with Open API (https://github.com/microsoft/semantic-kernel/tree/main/dotnet/samples/Concepts/Plugins). 

Open AI updated this blog post (https://openai.com/index/chatgpt-plugins/) noting "Note: OpenAI plugins have been deprecated. For more information about our products, check out [ChatGPT for everyone](https://openai.com/chatgpt/), [for Teams](https://openai.com/chatgpt/team/), and [for Enterprises](https://openai.com/chatgpt/enterprise/), plus our [APIs](https://openai.com/api/) for developers."

**Comment by raffaeler** [comment]
Hi @evchaki I am not sure we are talking about the same thing.
I specifically refer to local functions, not to OpenAI plugins (I saw that you can "import" the definitions inside a SK Plugin).
Local functions in the Azure OpenAI sdk v2 are defined using a json schema. You can create the function definition loading the json file.
My goal is to reuse the same json schema to define the SK plugin, in order to avoid creating two different adapters.

Thanks

**Reply by evchaki** [reply]
@markwallace-microsoft do you know if we have a sample for this?

**Comment by markwallace-microsoft** [comment]
@raffaeler 
The JSON schema we support for importing functions is the Open API schema, we standardised on this because there is widespread support. What is the JSON schema you are using? If it's something proprietary you could write you own import utilities or convert it to Open API format.

**Reply by raffaeler** [reply]
I am indeed using the OpenAI json schema.
The agents I am creating are deployed locally (with the running app) and not on the OpenAI side.
Basically I'd like to create a plugin, based on the JSON schema so that I can avoid using the SK specific attributes. This would allow me to share the agents with the codebase that is using the OpenAI SDK instead of the SK.

From what I understand from the specification, SK can only import definitions from the REST OpenAPI (Microsoft.SemanticKernel.Plugins.OpenApi.Extensions nuget package) and not from a straight JSON file that I have in my project. Is that correct?
If SK has an API to import the json directly, please let me know.

Thank you

**Reply by raffaeler** [reply]
I believe it's easy to get confused between OpenAPI (the standard for REST schema) and OpenAI who uses the JSON schema to define the Functions (aka Tools aka Agents aka Plugins in SK) that the LLM asks to invoke.

I hope this clarifies a bit my request.

**Reply by raffaeler** [reply]
Even better, here is an example of Function Definition (using the specs defined by OpenAI) for a Function/Tool/Agent that can be invoked by the LLM in the middle of a completion:
```json
{
    "name": "GetCurrentWeather",
    "description": "Get the current weather in a given location",
    "parameters": {
        "type": "object",
        "properties": {
            "location": {
                "type": "string",
                "description": "The city and state, e.g. San Francisco, CA"
            },
            "unit": {
                "type": "string",
                "enum": [ "celsius", "fahrenheit" ]
            }
        },
        "required": [ "location" ]
    }
}
```

My goal is to use this definition to define the Plugin in SK.
Thanks

**Comment by raffaeler** [comment]
The strategy used in the OpenAI function definitions looks like rudimental but it solves the problem of dependent libraries.
When I create a "plugin", I definitely don't want to depend on the SK libraries.
The same logic could potentially (and it is in my case) be used in other contexts where I don't want to add a dependency to SK.

To solve this, the cleanest and easiest solution is to:
- provide a way to register the functions (along with their metadata) without using interfaces or attributes that would involve a dependency to a package, but I did not find a way to do that. 

If there is a way to do that, I can parse the OpenAI json format by myself and register the function definitions.

@markwallace-microsoft could you please tell me if there is a way to imperatively build a plugin without using attributes or SK interfaces?

**Comment by raffaeler** [comment]
@markwallace-microsoft sorry for insisting, but you mentioned OpenAPI and I am instead talking  about OpenAI.
If you want to close this, just tell me whether I can programmatically create a SK Plugin from a configuration file or not.
If this is not the case I will have to write a generator for that.

Thanks

**Reply by markwallace-microsoft** [reply]
Hi @raffaeler, apologies for the delay in responding. We don't have support the OpenAI format at the moment. We do have a way to create KernelFunction's if you load the JSON yourself, see https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs#L30

Here's an example:
```csharp
var function = KernelFunctionFactory.CreateFromMethod((string location) =>
{
    functionCallCount++;
    return "Some weather";
}, "GetCurrentWeather");
var plugin = KernelPluginFactory.CreateFromFunctions("MyPlugin", [function]);
var openAIFunction = plugin.GetFunctionsMetadata().First().ToOpenAIFunction();
```

**Comment by raffaeler** [comment]
Thank you @markwallace-microsoft 
I did see it, but how can I specify the descriptions of the functions and the parameters?
I need to specify them manually as they are stored in the json using the OpenAI format.

**Comment by markwallace-microsoft** [comment, marked as answer]
@raffaeler 
You can deserialize the JSON into a model object that you provide and then convert that into the arguments required by `KernelPluginFactory.CreateFromFunctions`.

I tried this prompt

```
I have the following JSON  
{  
    "name": "GetCurrentWeather",  
    "description": "Get the current weather in a given location",  
    "parameters": {  
        "type": "object",  
        "properties": {  
            "location": {  
                "type": "string",  
                "description": "The city and state, e.g. San Francisco, CA"  
            },  
            "unit": {  
                "type": "string",  
                "enum": [ "celsius", "fahrenheit" ]  
            }  
        },  
        "required": [ "location" ]  
    }  
}  
Write the C# code to serialise this into a model class
```

And got this code (which should illustrate what you need to do)

Here's the C# code to create a model class and deserialize the JSON string into the class object using Newtonsoft.Json library:

First, you need to install the Newtonsoft.Json library, if not already installed. You can install it using NuGet package manager with the following command:

Install-Package Newtonsoft.Json  
 
2. Next, create the C# model classes to represent the JSON structure:


using Newtonsoft.Json;  
using System.Collections.Generic;  
  
public class GetCurrentWeather  
{  
    [JsonProperty("name")]  
    public string Name { get; set; }  
  
    [JsonProperty("description")]  
    public string Description { get; set; }  
  
    [JsonProperty("parameters")]  
    public Parameters Parameters { get; set; }  
}  
  
public class Parameters  
{  
    [JsonProperty("type")]  
    public string Type { get; set; }  
  
    [JsonProperty("properties")]  
    public Dictionary<string, Property> Properties { get; set; }  
  
    [JsonProperty("required")]  
    public List<string> Required { get; set; }  
}  
  
public class Property  
{  
    [JsonProperty("type")]  
    public string Type { get; set; }  
  
    [JsonProperty("description")]  
    public string Description { get; set; }  
  
    [JsonProperty("enum")]  
    public List<string> Enum { get; set; }  
}  
 
3. Deserialize the JSON string into the GetCurrentWeather class object:


using Newtonsoft.Json;  
  
string json = @"{  
    ""name"": ""GetCurrentWeather"",  
    ""description"": ""Get the current weather in a given location"",  
    ""parameters"": {  
        ""type"": ""object"",  
        ""properties"": {  
            ""location"": {  
                ""type"": ""string"",  
                ""description"": ""The city and state, e.g. San Francisco, CA""  
            },  
            ""unit"": {  
                ""type"": ""string"",  
                ""enum"": [ ""celsius"", ""fahrenheit"" ]  
            }  
        },  
        ""required"": [ ""location"" ]  
    }  
}";  
  
GetCurrentWeather currentWeather = JsonConvert.DeserializeObject<GetCurrentWeather>(json);  
 
Now you have deserialized the JSON into the GetCurrentWeather class, and you can use the currentWeather object to access the properties.

**Comment by raffaeler** [comment]
@markwallace-microsoft I see, thank you.
I mapped everything but the enums. Any idea how can I map the unit property?
Should I map it to an enum? This requires to create the enum type which is never used by my code, which is quite disappointing.
A "loosy" implementation would be more useful.

Thanks

**Reply by davidames** [reply]
You can now specify enum values using the Schema of KernelParameterMetadata.

Verified on AOAI by looking at the HTTP requests.


```
var schemaObject = new
{
    type = "string",
    @enum = enumValues,
    description = filter.PropertyDescription,
};


var schemaJson = JsonSerializer.Serialize(schemaObject);
var schema = KernelJsonSchema.Parse(schemaJson);

metadata.Add(new KernelParameterMetadata(filter.PropertyName)
{
    IsRequired = filter.IsRequired,
    Schema = schema
});    
```
